bftclient/src/bft_client.cpp

```diff
@@ -0,0 +1,162 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License"). You may not use
+// this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license
+// terms. Your use of these subcomponents is subject to the terms and conditions of the
+// subcomponent's license, as noted in the LICENSE file.
+
+#include "bftclient/bft_client.h"
+#include "bftengine/ClientMsgs.hpp"
+#include "assertUtils.hpp"
+
+namespace bft::client {
+
+// This function creates a ClientRequestMsg or a ClientPreProcessRequestMsg depending upon config.
+//
+// Since both of these are just instancds of a `ClientRequestMsgHeader` followed by the message
+// data, we construct them here, rather than relying on the type constructors embedded into the
+// bftEngine impl. This allows us to not have to link with the bftengine library, and also allows us
+// to return the messages as vectors with proper RAII based memory management.
+Msg MakeClientMsg(const RequestConfig& config, Msg&& request, bool read_only, uint16_t client_id) {
+  uint8_t flags = read_only ? READ_ONLY_REQ : EMPTY_FLAGS_REQ;
+  if (config.pre_execute) {
+    flags |= PRE_PROCESS_REQ;
+  }
+
+  auto header_size = sizeof(bftEngine::ClientRequestMsgHeader);
+
+  Msg msg(header_size + request.size());
+  bftEngine::ClientRequestMsgHeader* header = reinterpret_cast<bftEngine::ClientRequestMsgHeader*>(msg.data());
+  header->msgType = config.pre_execute ? PRE_PROCESS_REQUEST_MSG_TYPE : REQUEST_MSG_TYPE;
+  header->spanContextSize = 0;  // TODO (AJS): FIXME
+  header->idOfClientProxy = client_id;
+  header->flags = flags;
+  header->reqSeqNum = config.sequence_number;
+  header->requestLength = msg.size();
+  header->timeoutMilli = config.timeout.count();
+  header->cid_length = config.correlation_id.size();
+
+  // Copy the correlation ID
+  std::memcpy(msg.data() + header_size, config.correlation_id.data(), config.correlation_id.size());
+
+  // Copy the request data
+  std::memcpy(msg.data() + header_size + config.correlation_id.size(), request.data(), request.size());
+
+  return msg;
+}
+
+Reply Client::Send(const WriteConfig& config, Msg&& request) {
+  Assert(!outstanding_request_.has_value());
+  auto match_config = WriteConfigToMatchConfig(config);
+  bool read_only = false;
+  return Send(match_config, config.request, std::move(request), read_only);
+}
+
+Reply Client::Send(const ReadConfig& config, Msg&& request) {
+  Assert(!outstanding_request_.has_value());
+  auto match_config = ReadConfigToMatchConfig(config);
+  bool read_only = true;
+  return Send(match_config, config.request, std::move(request), read_only);
```

Comment:
I'm getting a feeling that division to ReadConfig/WriteConfig is artificial. Most of the code is the same for both.
Maybe the addition of a read_only flag to the configuration is enough.

Reply:
I don't feel it's artificiall. The semantics and capabilities of read and write requests are entirely different and the caller should be aware of how they work. The specific use case that drove this separation here is that Reads allow for more quorums than Writes. A read can be sent to just 1 or 2 replicas (via `All` or `MofN`), while a write cannot use these types of quorums. With this separation, we now get a *compile-time* check that the caller is appropriately using a quorum. If we had just one call, then we would need to write extra error handling code that would only be triggered at runtime. So we have both more code and less safety. We'd then also have to write tests for that error handling code rather than relying on correctness of the compiler.

 Furthermore, bool parameters are generally bad form, as you don't know what a literal true or false means at the call site and must look at the definition of the function.